### PR TITLE
CREATE/DROP SEQUENCEs

### PR DESCRIPTION
**Pull request by zianke**
[Design Document](https://github.com/adrianchang/terrier/wiki/Sequences-Design-Document)

### CODE REVIEW

**Comment by YingjingLu** [comment]
**Comments from Yingjing Lu** 
[yingjinl@andrew.cmu.edu](url)

**General Code Comments**
In general the code cpmplies with the cmake format style and is generally pretty straightforward to read. I would live to have a header at the database catalog to describe especially the ```CreateSequence()``` part what does the sequence do and any restrictions applied. There are some inline comments which is clear and concise. Would be better to have documentation header for some of the key functions APIs to denote the behavior and assumption for the functions. 

**Documentation** 
The document includes an overview with the key features listed by their SQL statements, which is quite straight forward. The description direct to the Postgres respective feature. Would be better to have further describe those feature behaviors for those who are not so familiar with Postgres as the students in this class. 

The scope lists all the modified files pretty comprehensive. If there are APIs exposes such as get sequences and execution functions, I would love to have some exposed and documented in the architectural design part for people who want to use this module in other parts of the implementation.

**Testing**
The testing are fairely comprehensive for the currently included featres in binder, parser, node planning, excution of sequence insertion, catalog in fetching the sequence. For all the above features that I inspected, the tests covers at least one of the testing instance as unittest. One thing that can definitely improve further is to add more autommatically generated tests to not just unit testing some enumerated instances but to dynamically test with multiple concurrent threads.

As documented in the potential problems, the pointers to sequence object and sequence schema remains undefined and is hardcoded in the testings. However, I believe that the team will solve this problem during the next phase of development.

File: src/catalog/database_catalog.cpp [hunk 1]
```diff
@@ -1,10 +1,11 @@
+#include "catalog/database_catalog.h"
```
Comment (parapluirevel) [inline src/catalog/database_catalog.cpp]: Optional: Keep #include statements in alphabetical order.

File: src/catalog/database_catalog.cpp [hunk 2]
```diff
@@ -1324,6 +1325,102 @@ std::vector<std::pair<common::ManagedPointer<storage::index::Index>, const Index
...
+  const auto name_varlen = storage::StorageUtil::CreateVarlen(name);
+
+  // Write the name into the PR
+  const auto name_offset = pg_class_all_cols_prm_[postgres::RELNAME_COL_OID];
+  auto *const name_ptr = class_insert_pr->AccessForceNotNull(name_offset);
+  *(reinterpret_cast<storage::VarlenEntry *>(name_ptr)) = name_varlen;
+
+  // Write the ns_oid into the PR
+  const auto ns_offset = pg_class_all_cols_prm_[postgres::RELNAMESPACE_COL_OID];
+  // TODO(zianke): Write the ns_oid into the PR
```
Comment (parapluirevel) [inline src/catalog/database_catalog.cpp]: Would be helpful to know what needs to happen (if anything) before you can write ns_old into the PR.

File: src/include/catalog/database_catalog.h [hunk 3]
```diff
@@ -248,6 +248,45 @@ class DatabaseCatalog {
...
+  bool CreateSequence(common::ManagedPointer<transaction::TransactionContext> txn, namespace_oid_t ns_oid,
+                      sequence_oid_t sequence_oid, const std::string &name);
+
+  /**
+   * Delete a sequence.
+   * @param txn for the operation
+   * @param sequence to be deleted
+   * @return true if the deletion succeeded, otherwise false.
+   */
+  bool DeleteSequence(common::ManagedPointer<transaction::TransactionContext> txn, sequence_oid_t sequence);
```
Comment (parapluirevel) [inline src/include/catalog/database_catalog.h:278]: Can declare helper method as private.

File: src/include/catalog/database_catalog.h [hunk 4]
```diff
@@ -248,6 +248,45 @@ class DatabaseCatalog {
   std::vector<std::pair<common::ManagedPointer<storage::index::Index>, const IndexSchema &>> GetIndexes(
       common::ManagedPointer<transaction::TransactionContext> txn, table_oid_t table);
 
+  /**
+   * Create the catalog entries for a new sequence.
```
Comment (parapluirevel) [inline src/include/catalog/database_catalog.h:252]: Optional: Maybe explain a little bit about these functions' preconditions/postconditions/error handling.

File: src/catalog/database_catalog.cpp [hunk 5]
```diff
@@ -1324,6 +1324,192 @@ std::vector<std::pair<common::ManagedPointer<storage::index::Index>, const Index
...
+
+  // Now we insert into indexes on pg_class
+  // Get PR initializers allocate a buffer from the largest one
+  const auto class_oid_index_init = classes_oid_index_->GetProjectedRowInitializer();
+  const auto class_name_index_init = classes_name_index_->GetProjectedRowInitializer();
+  const auto class_ns_index_init = classes_namespace_index_->GetProjectedRowInitializer();
+  TERRIER_ASSERT((class_name_index_init.ProjectedRowSize() >= class_oid_index_init.ProjectedRowSize()) &&
+                     (class_name_index_init.ProjectedRowSize() >= class_ns_index_init.ProjectedRowSize()),
+                 "Index buffer must be allocated based on the largest PR initializer");
+  auto *index_buffer = common::AllocationUtil::AllocateAligned(class_name_index_init.ProjectedRowSize());
```
Comment (parapluirevel) [inline src/catalog/database_catalog.cpp:1386]: Optional: maybe consider not using a raw pointer directly.

**Comment by wuwenw** [comment]
Comment from Wuwen Wang
wuwenw@andrew.cmu.edu

The code works fine locally and it is in a good format to be understood with provided comments. The overall quality of the code is good as it doesn’t have redundant code, and assert functions are used to ensure correctness at certain stages. I noticed that there is some commented code in database_catalog.cpp and it needs further implementation. 

Some parts of the code such as database_catalog.cpp and catalog_accessor only have inline comments but not headers. Though the function name is pretty interpretive, a header explains more about the functionality. At the plan node for creating sequence and dropping sequence, a third library of json, nlohmann is used and some functions are overridden without further documentation. It would help understanding if a line of comment explains the usage of the library. It is nice to see that there are many TODO comments in the code, and they give developers more guidance for future work.

Tests exist and they are pretty comprehensive, as they have tested the functionalities of parser, binder, optimizer, executor and catalog, including sequence creation and dropping, so it provides decent code coverage. As my teammate has mentioned , there are some minor defects with regard to the pointers, since they should not be hardcoded and checked in the tests.

File: src/catalog/database_catalog.cpp (hunk 5 above)
Comment (apavlo) [reply src/catalog/database_catalog.cpp:1412]: They are using the raw pointers correctly as it is allowed in this part of the code (catalog). This is not an issue.

**Comment by zianke** [comment]
Please refer to the [Final Code Drop](https://github.com/cmu-db/terrier/pull/897).

**Pull request closed without merging**